### PR TITLE
fix infinite loop in not fully packed bit-packed runs

### LINKED ISSUE #1458: Hang due to infinite loop when reading some parquet files with RLE encoding and bit packing
**Describe the bug**
Some parquet writers produce RLE encoded dictionaries with bit packing even if value count is not divisible by 8. 
This violates the assumption decoder makes about the number of values present in the bit-packed run, causing infinite loop and hang.

**To Reproduce**
Attempt to read parquet file where number of values is not power of 2 but RLE encoding with bit packing was used.

**Expected behavior**
File should be read correctly.

**Additional context**

### LINKED ISSUE #1548: Output of `ipc::reader::tests::projection_should_work` fails validation
**Describe the bug**
The `force_validate` check added in https://github.com/apache/arrow-rs/pull/1546 reveals that some array created as part of `projection_should_work` test fails validation. 

It is not clear if the validation is incorrect or something about ipc

**To Reproduce**
Uncomment the
```rust
    #[cfg(not(feature = "force_validate"))]
``` 

And then run the test like:
```shell
cargo test --features=force_validate -p arrow
```

lines for the following tests and the will fail:

```
---- ipc::reader::tests::projection_should_work stdout ----
thread 'ipc::reader::tests::projection_should_work' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidArgumentError("Last offset 251658240 of List(Field { name: \"item\", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: None }) is larger than values length 15")', arrow/src/array/data.rs:301:34

```

**Expected behavior**
The tests should pass when force_validate is on

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #1458 

Cherry-picked from #1460, will add test and then mark ready for review

# Rationale for this change
 
See ticket

# What changes are included in this PR?

See ticket

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by alamb** [comment]
fyi @anliakho2

File: parquet/src/encodings/rle.rs [hunk 1]
```diff
@@ -743,6 +753,42 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_truncated_rle() {
```
Comment (alamb) [inline parquet/src/encodings/rle.rs:757]: FWIW this test times out without the code in this PR 👍  which is a good sign to me that it covers the issue

```
test encodings::rle::tests::test_truncated_rle has been running for over 60 seconds
```

(I also tried removing each of the two cases -- dict and non dict and the test hung in both)

**Review by alamb — APPROVED** [review]
While I am not super familiar with the code, I did review the test and verified it fails without this code change. 

Also the code change makes sense to me. 

I also looked for other places that `get_batch` is called.

I found one in the DeltaBitPackedEncoder that maybe has a similar problem:


https://github.com/tustvold/arrow-rs/blob/fix-rle-infinite-loop/parquet/src/encodings/decoding.rs#L653

It also appears to be used a few other times. 

```
/Users/alamb/Software/arrow-rs/parquet/src/column/reader/decoder.rs
260:                 Ok(reader.get_batch::<i16>(&mut out[range], *bit_width as usize))
/Users/alamb/Software/arrow-rs/parquet/src/data_type.rs
696:             let values_read = bit_reader.get_batch(&mut buffer[..num_values], 1);
/Users/alamb/Software/arrow-rs/parquet/src/encodings/decoding.rs
655:                 .get_batch(&mut buffer[read..read + batch_to_read], bit_width);
/Users/alamb/Software/arrow-rs/parquet/src/encodings/levels.rs
271:                     decoder.get_batch::<i16>(&mut buffer[..len], bit_width as usize);
/Users/alamb/Software/arrow-rs/parquet/src/encodings/rle.rs
418:                 num_values = bit_reader.get_batch::<T>(
471:                     num_values = bit_reader.get_batch::<i32>(
/Users/alamb/Software/arrow-rs/parquet/src/util/bit_util.rs
1062:         let values_read = reader.get_batch::<T>(&mut batch, num_bits);
```

**Comment by tustvold** [comment]
> I found one in the DeltaBitPackedEncoder that maybe has a similar problem:

In this case the specification is very clear that the miniblock can't be truncated

> If there are not enough values to fill the last miniblock, we pad the miniblock so that its length is always the number of values in a full miniblock multiplied by the bit width. The values of the padding bits should be zero, but readers must accept paddings consisting of arbitrary bits as well.

And we return an error a line below if it is, so I think we should be ok.

Good shout to check though :+1: 

The other cases don't appear to have loops, and GenericColumnReader which drives them will bail out if a page is truncated

**Comment by alamb** [comment]
I plan to merge this after all (non miri) tests pass

**Pull request merged**